src/gui/MatchTemplatePanel.cpp

```diff
@@ -148,7 +148,7 @@ void MatchTemplatePanel::ResetDefaults( ) {
 
     if ( main_frame->current_project.is_open ) {
         ResumeRunCheckBox->SetValue(false);
-        if ( main_frame->current_project.database.ReturnNumberOfTemplateMatchingJobs( ) == 0 )
+        if ( match_template_results_panel->template_match_job_ids == NULL )
```

Comment:
I realize we have NULL all over the code base, but I would suggest we consider at least making new code follow a different approach, which seems to be well agreed upon behavior that avoids idiosyncratic bugs popping up with potential integer conversions etc.

Initialize the pointer to nullptr either
```cpp
// explicitly
template_match_job_ids = nullptr;
// or implicitly with brace initialization. 
template_match_job_ids{};

// Then just check the boolean value, not a comparision expression
if (match_template_results_panel->template_match_job_ids)
```

Reply:
I'm happy to go through and make the needed changes if you'd like, just lmk @jojoelfe !

Reply:
That's sounds absolutely like a better approach. I won't have a lot of time this week because daycare is on spring break and I would need some time to really understand it, so feel free to either push to this branch or create a new PR.

Reply:
@bHimes I went ahead and made the changes you suggested, and tested via the same gdb rbreak across all database.cpp functions, and there was no longer excessive database access. I also used force-push, as it was essentially the same changes, just with this nullptr method in mind, and I had already created a branch myself.

Just to clarify, I'm sure this would not normally be the desired method for updating a branch (especially when there's many changes), with preference instead being to create a local branch from the desired remote branch, pull from remote any time a new commit is added, and then simply add local commits on top to preserve the linear history? Just want to make sure I understand the general best practice rules when breaking them :)

Reply:
Yeah, not something you want to be doing.

For a branch you don't directly have access to you can:

```bash
# Check it out locally, where ID = pull request number and BRANCH_NAME is the branch someone else has been working on.
git fetch origin pull/ID/head:BRANCH_NAME
# Change to that branch and make edits
git switch BRANCH_NAME
# Then after editing
git push origin BRANCH_NAME 
# In this case, since the original branch was on JE's fork, I think this would just be a new branch and you would need a new PR. Otherwise, even if it is an existing branch, you wouldn't have to force push because you would be up to date when you fetched it.

```

Reply:
> For a branch you don't directly have access to you can:
> 
> ```shell
> # Check it out locally, where ID = pull request number and BRANCH_NAME is the branch someone else has been working on.
> git fetch origin pull/ID/head:BRANCH_NAME
> # Change to that branch and make edits
> git switch BRANCH_NAME
> # Then after editing
> git push origin BRANCH_NAME 

Got it, making note of this for future reference.

About this part though:

> ```shell
> # In this case, since the original branch was on JE's fork, I think this would just be a new branch and you would need a new PR. Otherwise, even if it is an existing branch, you wouldn't have to force push because you would be up to date when you fetched it.
> ```

1 question, and 1 thing I'm not sure of:

Without the force push, wouldn't I be adding a commit that simply switched from @jojoelfe 's solution to the more modern version you suggested? Or would it be better to undo to the commit, and push after pulling the branch, in this specific scenario? Basically, for this particular case, overwriting the initial commit was intentional since the change was a more robust format that was solving the same problem; but is it better to leave the commit history of such a change on the same branch?

And as for what I'm not sure of, it seems that this branch was from the main repo, as it shows up under the branches tab in this repo. This PR also displays the appropriate changes after the force push. I can definitely make a new PR though, if needed.

Reply:
No worries, and thanks for making the changes. I'll review them tomorrow. I just want to 2x check that we weren't using "NULL" in any weird ways that might break b/c of the change.

To your question: I think this is confusing b/c a PR isn't (afaik) a git thing; it's a github thing. Once a PR is opened, the downstream branch becomes available to the upstream repo (as read only) & github tracks all changes remotely.

So even thought it looks like BRANCH is in upstream it isn't. Pushing your local changes to upstream would then create a new branch (and require a new PR)

Alternatively, i think you could work it into the original branch, but I'm not sure. This is probably best to test out with a dummy repo.

See [the docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally?platform=linux#modifying-an-inactive-pull-request-locally) for what i'm referring to.

Reply:
for modifying the original branch, it looks like there are two options.

1) [fork the fork ](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork) which is what I've done in the past 

2) alternatively the PR creator may [enable push access](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) to the branch the or is based on

Reply:
Hey, I think for a simple case like this where there is only a small change that is completely reverted, the force-push kind of makes sense. The alternative would have been to open a new PR from a new branch and just close this one... If the changes are more complicated you might want to preserve the history and have commits that revert previous work.

Also this branch is in Tims repo, because that is what I end op doing for small things. Not sure why I am allowed to do that. In any case once a PR is opened, all project contributors can push to this branch even if it is in another fork. (One can disable this, but that's the default)

Reply:
I agree, for simple things maybe it feels like overkill, but i think a "do not overwrite anyone elses work" is probably a fair candidate for just being an absolute rule. TBF, i think in most cases, even force pushing on yourself is asking for trouble. Branches are cheap to create.

FWIW, I was able to use the method I suggested, modify and push back to this branch without any trouble. Maybe this was easy because the branch was in the repo?

Anyway, @jojoelfe can you take a look at my comment in this commit about memory leaks? I'm happy to fix with smart pointers, but I would like to be sure I'm not missing some garbarge collection somewhere.

Reply:
I don't really understand C++ garbage collection well enough. My general thought is that the AssetPanel gets only instanced once and never deleted as long as the GUI is running right? So garbage collection is not really that important...

Reply:
There is no garbage collection in the python/java sense - sorry, I meant "cleanup that we've implemented."

I guess the OS will clean up heap allocations when the GUI is closed then. Seems like bad practice, but I guess it has been working well enough...

Reply:
> @bHimes I went ahead and made the changes you suggested, and tested via the same gdb rbreak across all database.cpp functions, and there was no longer excessive database access.

@twagner9 would you mind running that test against the changes I made? Thanks!

Reply:
>@twagner9 would you mind running that test against the changes I made? Thanks!

The changes appear to work fine. I do encounter an odd shared library issue when running the rbreak across database.cpp, however: `Function "_ZN8Database25GetFromBatchSelect_NoCharIJPlEEEbDpT_" not defined in "/home/tim/VS_Projects/cisTEM/src/core/database.cpp".`

It seems to be likely that it's an issue with my lack of static linking on GCC; I spent a bit more time trying to do this; but ultimately abandoned the time sink. Otherwise, no database accesses seem to be occurring on mouse movement any longer.

Reply:
"Args" is just a keyword for the types in the template parameter pack. It shouldn't need to be defined anywhere. Is this at the link stage or while you are running things? There is always a chance there is an issue with instatiation since we compile the core library then link to it, in which case we could define (rather than just declare) GetFromBatchSElect_NoChar in the header.

I appreciate the extra info!

Reply:
I see, I haven't delved into template usage in C++ too much (courses that did during undergrad were in Java). This is while running, specifically while running in gdb and setting breakpoints across the entire database.cpp file. It still runs fine even past this, I guess it just doesn't set the breakpoint for that function because it doesn't get the definition, since it's not defined in the header? That seems to make sense to me.